patch_subset/cbor/cbor_utils.cc

```diff
@@ -0,0 +1,360 @@
+#include "patch_subset/cbor/cbor_utils.h"
+
+namespace patch_subset::cbor {
+
+using absl::string_view;
+
+StatusCode CborUtils::GetField(const cbor_item_t& cbor_map, int field_number,
+                               cbor_item_unique_ptr& out) {
+  if (!cbor_isa_map(&cbor_map) || field_number < 0) {
+    return StatusCode::kInvalidArgument;
+  }
+  size_t size = cbor_map_size(&cbor_map);
+  cbor_pair* pairs = cbor_map_handle(&cbor_map);
+  for (size_t i = 0; i < size; i++) {
+    cbor_pair* pair = pairs + i;
+    if (!cbor_is_int(pair->key)) {
+      // Ill formed map. Keys should all be ints. Skip this field.
+      continue;
+    }
+    int32_t id;
+    StatusCode sc = DecodeInt(*pair->key, &id);
+    if (sc != StatusCode::kOk) {
+      // Ill formed map. Keys should all be valid ints. Skip this field.
+      continue;
+    }
+    if (id == field_number) {
+      // Because we are putting the pointer into a cbor_item_unique_ptr, which
+      // will eventually call cbor_decref(), the item has a new "owner".
+      cbor_item_t* value = pair->value;
+      if (value == nullptr) {
+        return StatusCode::kInvalidArgument;
+      }
+      out.reset(cbor_incref(value));
+      return StatusCode::kOk;
+    }
+  }
+  // Field not found.
+  out.reset(nullptr);
+  return StatusCode::kNotFound;
+}
+
+StatusCode CborUtils::GetUInt64Field(const cbor_item_t& map, int field_number,
+                                     optional<uint64_t>& out) {
+  cbor_item_unique_ptr field = empty_cbor_ptr();
+  StatusCode sc = GetField(map, field_number, field);
+  if (sc == StatusCode::kNotFound) {
+    out.reset();
+    return StatusCode::kOk;
+  } else if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  uint64_t result;
+  sc = DecodeUInt64(*field, &result);
+  if (sc != StatusCode::kOk) {
+    return sc;
+  }
+  out.emplace(result);
+  return StatusCode::kOk;
+}
+
+// TODO: This method could ignore failures, returning "", if we want.
+StatusCode CborUtils::GetStringField(const cbor_item_t& map, int field_number,
+                                     optional<string>& out) {
```

Comment:
We don't use the cbor string type anywhere currently in the spec, we only use the byte string type. So this method can be dropped.

Reply:
and the SetStringField, DecodeString, EncodeString.

Reply:
It is used, in the next-next CL.

Reply:
Just curious what is it needed for? Since we don't use the string type anywhere in the spec.